src/lib/Simplify.hs

```diff
@@ -258,7 +317,16 @@ simplifyApp f xs =
             naryApp a' (map sink $ toList xs)
         caseExpr <- caseComputingEffs e alts' resultTy
         dropSubst $ simplifyExpr caseExpr
-      _ -> naryApp atom $ toList xs
+      Var v ->
+        lookupAtomName v >>= \case
+          TopFunBound _ (UnspecializedTopFun numSpecializationArgs _) ->
```

Comment:
Should this have a case for `TopFunBound _ (SpecializedTopFun _ _)`?  If not, add a comment for why not?

Reply:
It does have that case, it's just not a special one. It's just handled by the fallback `_` pattern and emits an application.